lib/salus/repo_searcher.rb

```diff
@@ -0,0 +1,201 @@
+      @scanner_config.key?('recursion')
+    end
+
+    ##
+    # @return [Array<String>] List of directories from the recusion.directories
+    # configuration prefixed to be relative to @path_to_repo
+    #
+    def static_directories
+      dirs = @scanner_config.dig('recursion', 'directories') || []
+      resolve_dirs(dirs)
+    end
+
+    ##
+    # @return [Array<String>] List of static files from the recursion.static_files
+    # configuration
+    def static_files
+      @static_files ||= @scanner_config.dig('recursion', 'static_files') || []
+      @static_files
+    end
+
+    ##
+    # @return [Array<String>] List of directories that match the rules defined
+    # in recusion.directories_matching after searching. Prefixed to be relative
+    # to @path_to_repo
+    def dynamic_directories
+      matching_rules = @scanner_config.dig('recursion', 'directories_matching') || []
+      # Let's search for any dirs matching these rules
+      directories_matching(matching_rules)
+    end
+
+    ##
+    # @param [Array<String>] dirs List of directories
+    # @return [Array<String>] Filtered set of directories after removing
+    # directories that start with any directories listed in the
+    # recursion.directory_exclusions configuration
+    def filter_out_exlcusions(dirs)
+      return [] if dirs.nil?
+
+      exclusions = @scanner_config.dig('recursion', 'directory_exclusions') || []
+      exclusions.map! { |dir| File.join(@path_to_repo.to_s, dir) }
+      return dirs if dirs.empty? || exclusions.empty?
+
+      dirs.reject { |dir| dir.start_with?(*exclusions) }
+    end
+
+    ##
+    # @param [String] filename Filename to search for.
+    # @return [Array<String>] List of files found to have the passed
+    # filename.  Recusive search from @path_to_repo is preformed to find
+    # the mathces
+    def search_files_named(filename)
+      run_rg("rg", "--files", "-g", filename)
+    end
+
+    ##
+    # @param [String] filename Filename to search for.
+    # @param [String] content File content to search for.
+    # @return [Array<String>] List of files found to contain the passed
+    # content and matching the provided filename.  Recusive search from
+    # @path_to_repo is preformed to find the mathces
+    def search_files_named_containing(filename, content)
+      run_rg("rg", "--files-with-matches", content, "--glob", filename)
+    end
+
+    ##
+    # @param [String] content File content to search for.
+    # @return [Array<String>] List of files found to contain the passed
+    # content.  Recusive search from @path_to_repo is preformed to find
+    # the mathces
+    def search_files_containing(content)
+      run_rg("rg", "-l", content)
+    end
+
+    ##
+    # @param [Array] args CLI rg command and options to run
+    # @return [Array<String>] Relative path of files returned
+    # from running the command.
+    def run_rg(*args)
+      data = nil
+      Dir.chdir(@path_to_repo) do
+        data = IO.popen(args).read
```

Comment:
What if `rg` throws exception or outputs something unexpected?

Reply:
We will fail - open to thoughts, if we are unable to scan the expected / configured directories I'm okay with a fail.